### PR TITLE
[New3D] Support foxglove.PointCloud

### PR DESCRIPTION
**Pull request by jhurliman**
**User-Facing Changes**

- Adds support for `foxglove.PointCloud` and `foxglove.FrameTransform` messages to `3D (Beta)` panel

**Description**

Adds support for these messages and storybooks.

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/Renderer.ts [hunk 1]
```diff
@@ -550,7 +558,17 @@ export class Renderer extends EventEmitter<RendererEvents> {
       this.addCoordinateFrame(frameId);
     }
 
-    if (TF_DATATYPES.has(datatype)) {
+    // If this message has a top-level frame_id, scrape it
+    const maybeHasFrameId = message as Partial<{ frame_id: string }>;
+    if (maybeHasFrameId.frame_id != undefined) {
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/Renderer.ts:563]:
```suggestion
    if (typeof maybeHasFrameId.frame_id === "string") {
```

File: packages/studio-base/src/panels/ThreeDeeRender/foxglove.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+export const FRAME_TRANSFORM_DATATYPES = new Set<string>();
+addFoxgloveDataType(FRAME_TRANSFORM_DATATYPES, "foxglove.FrameTransform");
+
+export const POINTCLOUD_DATATYPES = new Set<string>();
+addFoxgloveDataType(POINTCLOUD_DATATYPES, "foxglove.PointCloud");
+
+// Expand a single Foxglove dataType into variations for ROS1 and ROS2 then add
+// them to the given output set
+function addFoxgloveDataType(output: Set<string>, dataType: string): Set<string> {
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/foxglove.ts:13]:
```suggestion
function addFoxgloveSchema(output: Set<string>, schemaName: string): Set<string> {
```

File: packages/studio-base/src/panels/ThreeDeeRender/foxglove.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  // Add the ROS1 variation: foxglove/PointCloud
+  output.add(`${dataType}`.replace(/\./g, "/"));
+
+  // Add the ROS2 variation: foxglove/msg/PointCloud
+  const parts = dataType.split(".");
+  if (parts.length > 1) {
+    const base = parts[0];
+    const leaf = parts.slice(1).join("/");
+    output.add(`${base}/msg/${leaf}`);
+  }
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/foxglove.ts:26]: These are wrong, they should be `foxglove_msgs/*`

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/pointClouds/fieldReaders.ts [hunk 4]
```diff
@@ -39,11 +41,15 @@ export function float64Reader(fieldOffset: number): FieldReader {
...
+  field: PackedElementField | PointField,
   pointStep: number,
   forceType?: PointFieldType,
 ): FieldReader | undefined {
-  switch (forceType ?? field.datatype) {
+  switch (
+    forceType ??
+    (field as Partial<PackedElementField>).type ??
+    (field as PointField).datatype
+  ) {
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/pointClouds/fieldReaders.ts:52]: These values are wrong, you need to use the PackedElementField enum exported from here https://github.com/foxglove/schemas/blob/main/schemas/typescript/NumericType.ts

Note that we already have: https://github.com/foxglove/studio/blob/480b474f8386438939ce44cd0d26c9f005ebeba8/packages/studio-base/src/panels/ThreeDimensionalViz/normalizeMessages.ts#L104

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/PointCloudsAndLaserScans.ts [hunk 5]
```diff
@@ -284,7 +298,78 @@ export class PointCloudsAndLaserScans extends SceneExtension<PointCloudAndLaserS
     }
   };
 
-  handlePointCloud = (messageEvent: PartialMessageEvent<PointCloud2>): void => {
+  handleFgPointCloud = (messageEvent: PartialMessageEvent<PointCloud>): void => {
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/PointCloudsAndLaserScans.ts:301]:
```suggestion
  handleFoxglovePointCloud = (messageEvent: PartialMessageEvent<PointCloud>): void => {
```

Comment (jtbandes) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/PointCloudsAndLaserScans.ts:301]: Do we really need another method for this, or could we use the normalize function to make all point clouds go through one code path? Note that we already have a conversion function at https://github.com/foxglove/studio/blob/480b474f8386438939ce44cd0d26c9f005ebeba8/packages/studio-base/src/panels/ThreeDimensionalViz/normalizeMessages.ts#L125

Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/PointCloudsAndLaserScans.ts:301]: I thought about this but it seemed like a larger refactor to track both the actual original message (for picking / linked variables) and the transformed message. And this way we avoid creating another allocation for this message and each PackedElementField.

**Review by jtbandes — APPROVED** [review]

**Pull request merged**